sklearn/tests/test_multioutput.py

```diff
@@ -162,6 +163,7 @@ def test_multi_target_sample_weights():
 classes = list(map(np.unique, (y1, y2, y3)))
 
 
+@fails_if_pypy  # FIXME
```

Comment:
This fails due to not being able to allocate memory somewhere in loky,
<details>

```
    def test_multi_output_classification_partial_fit_parallelism():
        sgd_linear_clf = SGDClassifier(loss='log', random_state=1, max_iter=5)
        mor = MultiOutputClassifier(sgd_linear_clf, n_jobs=-1)
>       mor.partial_fit(X, y, classes)

sklearn/tests/test_multioutput.py:168: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
sklearn/utils/metaestimators.py:118: in <lambda>
    out = lambda *args, **kwargs: self.fn(obj, *args, **kwargs)
sklearn/multioutput.py:121: in partial_fit
    sample_weight, first_time) for i in range(y.shape[1]))
sklearn/externals/joblib/parallel.py:996: in __call__
    self.retrieve()
sklearn/externals/joblib/parallel.py:899: in retrieve
    self._output.extend(job.get(timeout=self.timeout))
pypy-env/lib_pypy/_functools.py:80: in __call__
    return self._func(*(self._args + fargs), **fkeywords)
sklearn/externals/joblib/_parallel_backends.py:517: in wrap_future_result
    return future.result(timeout=timeout)
/usr/local/lib-python/3/concurrent/futures/_base.py:405: in result
    return self.__get_result()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Future at 0x10052410 state=finished raised BrokenProcessPool>

    def __get_result(self):
        if self._exception:
>           raise self._exception
E           sklearn.externals.joblib.externals.loky.process_executor.BrokenProcessPool: A task has failed to un-serialize. Please ensure that the arguments of the function are all picklable.

/usr/local/lib-python/3/concurrent/futures/_base.py:357: BrokenProcessPool
----------------------------- Captured stderr call -----------------------------
/usr/local/lib-python/3/multiprocessing/semaphore_tracker.py:129: UserWarning: semaphore_tracker: There appear to be 1 leaked semaphores to clean up at shutdown
  len(cache))
Traceback (most recent call last):
  File "/usr/local/lib-python/3/runpy.py", line 183, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
  File "/usr/local/lib-python/3/runpy.py", line 109, in _get_module_details
    __import__(pkg_name)
  File "/root/project/sklearn/__init__.py", line 64, in <module>
    from .base import clone
  File "/root/project/sklearn/base.py", line 13, in <module>
    from .utils.fixes import signature
  File "/root/project/sklearn/utils/__init__.py", line 13, in <module>
    from .validation import (as_float_array,
  File "/root/project/sklearn/utils/validation.py", line 27, in <module>
    from ..utils._joblib import Memory
  File "/root/project/sklearn/utils/_joblib.py", line 18, in <module>
    from ..externals.joblib import __all__   # noqa
  File "/root/project/sklearn/externals/joblib/__init__.py", line 119, in <module>
    from .parallel import Parallel
  File "/root/project/sklearn/externals/joblib/parallel.py", line 28, in <module>
    from ._parallel_backends import (FallbackToBackend, MultiprocessingBackend,
  File "/root/project/sklearn/externals/joblib/_parallel_backends.py", line 22, in <module>
    from .executor import get_memmapping_executor
  File "/root/project/sklearn/externals/joblib/executor.py", line 14, in <module>
    from .externals.loky.reusable_executor import get_reusable_executor
  File "/root/project/sklearn/externals/joblib/externals/loky/__init__.py", line 11, in <module>
    from .backend.context import cpu_count
OSError: [Errno 12] Cannot allocate memory
Traceback (most recent call last):
  File "/usr/local/lib-python/3/runpy.py", line 183, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
  File "/usr/local/lib-python/3/runpy.py", line 109, in _get_module_details
    __import__(pkg_name)
  File "/root/project/sklearn/__init__.py", line 64, in <module>
    from .base import clone
  File "/root/project/sklearn/base.py", line 13, in <module>
    from .utils.fixes import signature
  File "/root/project/sklearn/utils/__init__.py", line 13, in <module>
    from .validation import (as_float_array,
  File "/root/project/sklearn/utils/validation.py", line 27, in <module>
    from ..utils._joblib import Memory
  File "/root/project/sklearn/utils/_joblib.py", line 18, in <module>
    from ..externals.joblib import __all__   # noqa
  File "/root/project/sklearn/externals/joblib/__init__.py", line 119, in <module>
    from .parallel import Parallel
  File "/root/project/sklearn/externals/joblib/parallel.py", line 28, in <module>
    from ._parallel_backends import (FallbackToBackend, MultiprocessingBackend,
  File "/root/project/sklearn/externals/joblib/_parallel_backends.py", line 20, in <module>
    from .pool import MemmappingPool
  File "/root/project/sklearn/externals/joblib/pool.py", line 33, in <module>
    from ._memmapping_reducer import get_memmapping_reducers
OSError: [Errno 12] Cannot allocate memory
Traceback (most recent call last):
  File "/usr/local/lib-python/3/runpy.py", line 183, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
  File "/usr/local/lib-python/3/runpy.py", line 109, in _get_module_details
    __import__(pkg_name)
  File "/root/project/sklearn/__init__.py", line 64, in <module>
    from .base import clone
  File "/root/project/sklearn/base.py", line 13, in <module>
    from .utils.fixes import signature
  File "/root/project/sklearn/utils/__init__.py", line 13, in <module>
    from .validation import (as_float_array,
  File "/root/project/sklearn/utils/validation.py", line 27, in <module>
    from ..utils._joblib import Memory
  File "/root/project/sklearn/utils/_joblib.py", line 18, in <module>
    from ..externals.joblib import __all__   # noqa
  File "/root/project/sklearn/externals/joblib/__init__.py", line 119, in <module>
    from .parallel import Parallel
  File "/root/project/sklearn/externals/joblib/parallel.py", line 28, in <module>
    from ._parallel_backends import (FallbackToBackend, MultiprocessingBackend,
  File "/root/project/sklearn/externals/joblib/_parallel_backends.py", line 22, in <module>
    from .executor import get_memmapping_executor
  File "/root/project/sklearn/externals/joblib/executor.py", line 14, in <module>
    from .externals.loky.reusable_executor import get_reusable_executor
  File "/root/project/sklearn/externals/joblib/externals/loky/__init__.py", line 12, in <module>
    from .reusable_executor import get_reusable_executor
  File "/root/project/sklearn/externals/joblib/externals/loky/reusable_executor.py", line 11, in <module>
    from .process_executor import ProcessPoolExecutor, EXTRA_QUEUED_CALLS
  File "/root/project/sklearn/externals/joblib/externals/loky/process_executor.py", line 82, in <module>
    from .backend.utils import recursive_terminate
OSError: [Errno 12] Cannot allocate memory
```
</details>

will report this upstream.

Reply:
I don't know if it's the case on circle but `n_jobs=-1` is dangerous on travis with pypy: the robust detection of the current number of cpu based on `os.sched_getaffinity` is not possible in pypy (not implemented). Instead joblib / loky fallsback to the basic cpu count detection that can report 32 cores on travis while on 2 are actually available.

Launching 32 pypy worker processes might exhaust the available memory of the worker.

Two solutions:

- `export LOKY_MAX_CPU_COUNT="2"` in the pypy test setup
- change the test itself to `n_jobs=2`

I think I prefer the `LOKY_MAX_CPU_COUNT` solution.